extensions/amp-lightbox/1.0/lightbox.js

```diff
@@ -0,0 +1,196 @@
+  'fly-in-top': [
+    {opacity: 0, transform: 'translate(0,-100%)', visibility: 'visible'},
+    {opacity: 1, transform: 'translate(0, 0)', visibility: 'visible'},
+  ],
+  'fly-in-bottom': [
+    {opacity: 0, transform: 'translate(0, 100%)', visibility: 'visible'},
+    {opacity: 1, transform: 'translate(0, 0)', visibility: 'visible'},
+  ],
+};
+
+/**
+ * @param {string} selector
+ * @param {number} time
+ * @return {number}
+ */
+function getElement(selector, time = 5) {
+  return new Promise(async (resolve, reject) => {
+    if (time <= 0) {
+      return reject(`${selector} not found`);
+    }
+    const elements = document.querySelectorAll(selector);
+    if (elements.length) {
+      return resolve(elements);
+    }
+    return setTimeout(async () => await getElement(selector, time - 1), 1000);
+  });
+}
+
+/**
+ * @param {!LightboxDef.Props} props
+ * @param {{current: (!LightboxDef.LightboxApi|null)}} ref
+ * @return {PreactDef.Renderable}
+ */
+function LightboxWithRef(
+  {
+    id,
+    animateIn = 'fade-in',
+    closeButtonAriaLabel,
+    // eslint-disable-next-line no-unused-vars
+    scrollable, // (TODO: discussion)
+    children,
+    onOpen,
+    initialOpen,
+    ...rest
+  },
+  ref
+) {
+  const [show, setShow] = useState(initialOpen);
+  const [visible, setVisible] = useState(false);
+  const styleClasses = useStyles();
+  const onOpenFnRef = useRef();
+  onOpenFnRef.current = onOpen;
+  const animateInRef = useRef();
+  animateInRef.current = animateIn;
+  const lightboxContainWrapperRef = useRef();
+  lightboxContainWrapperRef.current = `.${styleClasses.lightboxContainWrapper}`;
+
+  useImperativeHandle(
+    ref,
+    () => ({
+      open: () => {
+        setShow(true);
+      },
+      close: () => {
+        setShow(false);
+      },
+    }),
+    []
+  );
+
+  useEffect(() => {
+    if (show) {
+      setVisible(true);
+    } else {
+      getElement(lightboxContainWrapperRef.current)
+        .then((element) => {
+          element = element[0];
+          const animation = element
+            .animate(ANIMATION_PRESETS[animateInRef.current], {
+              duration: ANIMATION_DURATION,
+              direction: 'reverse',
+              fill: 'both',
+            })
+            .finished.then((opt_el) => {
+              setVisible(false);
+            })
+            .catch((err) => {
+              throw Error('Lightbox closing animation failed', err);
+            });
+          return () => animation.cancel();
+        })
+        .catch((err) => {
+          throw Error(err);
+        });
+    }
+  }, [show]);
+
+  useEffect(() => {
+    if (!show && !visible) {
+      return;
+    }
+    if (visible) {
+      getElement(lightboxContainWrapperRef.current)
+        .then((element) => {
+          element = element[0];
+          const animation = element.animate(
+            ANIMATION_PRESETS[animateInRef.current],
+            {
+              duration: ANIMATION_DURATION,
+              fill: 'both',
```

Comment:
While `fill:both` here?

Reply:
Fill both is for the persistence of styling, I found that it was needed to stop flickering

Reply:
That's ok as long as it's not used as the definitive `finally` clause. Our code should ensure that the styles are as expected in the end. There are also some dangers that `fill: both` has the highest specificity of styles and sometimes can cause conflicts.